platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterRetriever.java

```diff
@@ -0,0 +1,180 @@
+        final long round = getRound(state);
+        final ReadableSingletonState<RosterState> rosterState =
+                state.getReadableStates("RosterServiceImpl").getSingleton("ROSTER_STATE");
+        // replace with binary search when/if the list size becomes unreasonably large (100s of entries or more)
+        final List<RoundRosterPair> roundRosterPairs = rosterState.get().roundRosterPairs();
+        for (int i = 0; i < roundRosterPairs.size(); i++) {
+            final RoundRosterPair roundRosterPair = roundRosterPairs.get(i);
+            if (roundRosterPair.roundNumber() <= round) {
+                return roundRosterPair.activeRosterHash();
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Get the current round of the given state.
+     *
+     * @param state a state
+     * @return the round of the state
+     */
+    public static long getRound(@NonNull final State state) {
+        // TODO: update to States API after https://github.com/hashgraph/hedera-services/pull/14551 :
+        final MerkleRoot stateRoot = (MerkleRoot) state;
+        return stateRoot.getPlatformState().getRound();
+    }
+
+    /**
+     * Builds a Roster object out of a given AddressBook object.
+     *
+     * @param addressBook an AddressBook
+     * @return a Roster
+     */
+    @NonNull
+    public static Roster buildRoster(@NonNull final AddressBook addressBook) {
+        return Roster.newBuilder()
+                .rosters(addressBook.getNodeIdSet().stream()
+                        .map(addressBook::getAddress)
+                        .map(address -> {
+                            try {
+                                return RosterEntry.newBuilder()
+                                        .nodeId(address.getNodeId().id())
+                                        .weight(address.getWeight())
+                                        .gossipCaCertificate(
+                                                Bytes.wrap(address.getSigCert().getTBSCertificate()))
+                                        .gossipEndpoint(List.of(
+                                                        Pair.of(
+                                                                address.getHostnameExternal(),
+                                                                address.getPortExternal()),
+                                                        Pair.of(
+                                                                address.getHostnameInternal(),
+                                                                address.getPortInternal()))
+                                                .stream()
+                                                .filter(pair -> pair.left() != null
+                                                        && !pair.left().isBlank()
+                                                        && pair.right() != 0)
+                                                .distinct()
```

Comment:
do we really need a distinct call here? Seems like an overkill.

Reply:
Yes, I overheard somewhere (I don't recall the reference) that the existing AddressBook may in fact list the same address for both the internal and external endpoints, so I decided to remove any duplicates just in case. This doesn't seem to have a potential to create any issues, does it?

Also, the expectation is that the platform will migrate over to Rosters after which (a couple of releases or so?) this legacy code path that converts an AddressBook to a Roster could be removed altogether.

Let me know what you think.

Reply:
Some considerations i have here are:
1. The API chaining here seems to be handful. Is it worth breaking this method down for clarity?
2. There are max 2 endpoints per address. If true, the stream, filter, distinct, and map calls are operating on two elements? Is that correct?

Reply:
This isn't a two-liner to be sure, but it looks manageable. I'd like to re-iterate that this method will be removed very soon - as soon as we replace the AddressBook with the Roster, which should be completed with the TSS-Roster project as a whole. So I'd prefer to keep it as is, in as little number of methods as possible.

And yes, that stream operates on a maximum of 2 elements, one of which is likely a duplicate of another.

Reply:
ok. If it's going to go away soon, maybe that's fine.